Add realtime inference consumer

yatharthranjan opened the pull request:
Realtime Consumers are consumers for the results from the invocation of a model from the realtime analysis pipeline (see [model-builder](https://github.com/RADAR-base/model-builder/tree/dev/model-builder), [model-invocation-endpoint](https://github.com/RADAR-base/model-builder/tree/dev/model-invocation-endpoint)and [ksqldb](https://github.com/RADAR-base/ksql-extras)). The results from the pipeline are in JSON and the invocation is called by a [KSQL function](https://github.com/RADAR-base/ksql-extras/blob/main/src/main/java/org/radarbase/ksql/udf/RestInferenceUdf.java) which gets inference results then puts the results in a kafka topic. These consumers are meant to consume data from this topic.

3 terms used -

1. `Action`: Any task to be performed is an action. This can be sending a notification or an email
   or just logging something.
2. `Condition`: This is a predicate that will be evaluated on the incoming data. For instance, this
   can be used to evaluate if the ML model run crossed a threshold of some sort.
3. `RealtimeInferenceConsumer`: This is the base consumer for realtime results topics. Each consumer
   can be configured with a set of Conditions and Actions. If all the conditions evaluate to be
   true, then all the actions are triggered.

Currently, only one Condition is provided which evaluates a [JSONPath](https://github.com/json-path/JsonPath) expression. This should be sufficient for most simple use cases. 
More complex use cases can create concrete implementations of the `Condition`
interface or `ConditionBase` abstract class. Currently, the JsonPath expression is read from the
configuration file and hence is static for a particular consumer and topic. Later, we can also
provide this through AppConfig so it can be dynamic based on userId or projectId.

The supported config (for instance an intervention using aRMT app) is of the format -

```yaml
realtime-consumers:
  - name: 'lstm-lung-study-consumer' # Name of the consumer
    topic: 'lung_study_lstm_ad_inference' # Kafka topic to consume from (this should contain the inference results)
    notify_errors:
      email_addresses:
        - 'admin@radar-base.org'
    conditions:
      - name: 'LocalJsonPathCondition' # Name of the condition
        projects: [ 'radar-test' ]  # Only evaluate for these projects
        subjects: [ 'sub-1', 'sub-2' ]  # Only evaluate for these subjects
        properties:
          jsonpath: '$[?(@.invocation_result.anomaly_detected == true)]' # JsonPath expression to evaluate
          key: 'invocation_result' # Key that contains data to evaluate 
    actions:
      - name: 'ActiveAppNotificationAction' # Name of the action
        projects: [ 'radar-test' ] # Only execute for these projects
        subjects: [ 'sub-1', 'sub-2' ]  # Only execute for these subjects
        properties:
          questionnaire_name: 'ers' # Name of the questionnaire to trigger
          time_of_day: '09:00:00' # Local user time of day to trigger at
          default_timezone: 'Europe/London' # Default timezone to use for the time of day if not found in the appserver
          appserver_base_url: 'http://localhost:8080/' # Base URL of the appserver
          management_portal_token_url: 'http://localhost/managementportal/api/oauth/token' # URL to get the management portal token
          client_id: 'realtime_consumer' # Client ID for the management portal
          client_secret: 'secret' # Client secret for the management portal
          metadata_key: 'invocation_result' # Key that contains the metdata to be forwarded to the aRMT app
      - name: 'EmailUserAction'
        projects: [ 'radar-test' ]
        subjects: [ 'sub-1', 'sub-2' ]
        properties:
          email_addresses: [ 'admin@radar-base.org' ]
```

Note: The `properties` section is specific to each Action and Condition. Please take a look at the
condition and action docs for the keys supported. If the `projects` or `subjects` key is not
specified the action and condition will be used on all projects or subjects respectively.

yatharthranjan committed:
Merge pull request #77 from RADAR-base/release-0.3.0

Release 0.3.0

nivemaham committed:
Merge pull request #83 from RADAR-base/release-0.4.0

Release 0.4.0

yatharthranjan committed:
Initial impl for realtime consumers

yatharthranjan committed:
Fix dependencies

yatharthranjan committed:
First complete set of conditions and actions

- Includes various implementations includeing appserver notification actions and email user.
- Conditions evaluation using json path (currently locally, later read jsonpath from appconfig too).

yatharthranjan committed:
better use of enums

yatharthranjan committed:
add basic javadocs

yatharthranjan committed:
simplify defaults and enum

blootsvoets commented on src/main/java/org/radarcns/consumer/realtime/action/ActiveAppNotificationAction.java:
```diff
@@ -0,0 +1,135 @@
+                .getProperties()
+                .getOrDefault(
+                    "management_portal_token_url",
+                    "https://radar-cns-platform.rosalind.kcl.ac.uk/managementportal/api/ouath/token");
```
Preferably keep this out of the code. If you want to provide defaults, it would be better to define a base URL somewhere, and use that plus the default path as the default.

blootsvoets commented on src/main/java/org/radarcns/consumer/realtime/action/EmailUserAction.java:
```diff
@@ -0,0 +1,78 @@
+        new EmailSender(
+            (String) props.getOrDefault("host", "localhost"),
+            (Integer) props.getOrDefault("port", 25),
+            (String) props.getOrDefault("from", "admin@radarbase.org"),
```
This should not have a default.

blootsvoets commented on src/main/java/org/radarcns/consumer/realtime/action/appserver/AppserverClient.java:
```diff
@@ -0,0 +1,117 @@
+    objectMapper = new ObjectMapper();
+
+    if (tokenUrl == null || tokenUrl.isEmpty()) {
+      logger.warn("MP Token url was not provided. Will use unauthenticated requests to appserver.");
```
Is this ever the case? It would be better to exclude this.

yatharthranjan replied:
Unlikely in a production environment. The appserver does support a no-security mode where unauthenticated requests can be made (particularly useful for dev environments). I will remove it once I have tested it with the appserver.

blootsvoets reviewed (commented):
Yes this looks like a very good start! At some point the threshold may need to come from radar-app-config service instead of from a fixed value. This may also be a good entry point to enable / disable the condition for a given project (e.g. to only process records for certain projects).

yatharthranjan committed:
add support for filtering conditions and actions based on projects.

yatharthranjan committed:
remove unintended check

nivemaham committed:
add appserver related components to docker-compose.yml file

nivemaham commented on src/main/java/org/radarcns/consumer/realtime/RealtimeInferenceConsumer.java:
```diff
@@ -0,0 +1,185 @@
+ * <p>To be used with the model-invocation-endpoint and KSQL API_INFERENCE function to evaluate and
+ * take action on incoming results from realtime inference on data.
+ */
+public class RealtimeInferenceConsumer implements KafkaMonitor {
```
If I understand it correctly, all of these Consumers are realtime. We can remove that word from the class name I think.
Why not reuse/extend from existing AbstractKafkaMonitor?

yatharthranjan replied:
I think the adjective realtime was added for inference and not for consumers. So these are consumers for realtime inference as compared to batch or historic inference. But I agree if it seems confusing best to remove it. AbstractKafkaMonitor was not extended just to allow more flexibility (like for now a different constructor spec) but can definitely extend from AbstractKafkaMonitor if we need functionality already present there.

nivemaham commented on src/main/java/org/radarcns/consumer/realtime/RealtimeInferenceConsumerFactory.java:
```diff
@@ -0,0 +1,30 @@
+            .map(c -> createConsumer(handler.getRadarProperties(), c)));
+  }
+
+  private static KafkaMonitor createConsumer(
```
I would call it InferenceMonitor, InferenceMonitorFactory for consistency

nivemaham commented on src/main/java/org/radarcns/consumer/realtime/action/appserver/AppserverClient.java:
```diff
@@ -0,0 +1,117 @@
+    }
+  }
+
+  public Map<String, Object> createMessage(
```
you are referring to the create operation that is done on the service that receives the request. Right? I think this should be either sendMessage or addMessage.

yatharthranjan replied:
yes good point

yatharthranjan committed:
Merge remote-tracking branch 'origin/master' into add_realtime_consumer

yatharthranjan committed:
minor fixes to make build successful

nivemaham committed:
Merge branch 'add_realtime_consumer' into intervention-trigger

nivemaham committed:
resolve conflicts

nivemaham committed:
move monitor configs to separate package

nivemaham committed:
rename NotifyConfig.java as EmailNotifyConfig

nivemaham committed:
restructure email server config.

nivemaham committed:
new intervention config model

nivemaham committed:
Commit InterventionMonitor

nivemaham committed:
Commit config format and object creation

blootsvoets committed:
Implemented intervention monitor and updated build config

Also replaced all deprecated code

blootsvoets committed:
Simplify docker build to remove rest proxy

blootsvoets committed:
Fix tests

blootsvoets committed:
Fix docker setup

blootsvoets committed:
Use sh in docker

blootsvoets committed:
Move email senders and simplify config

blootsvoets committed:
Tested and fixed tests

blootsvoets committed:
Tested full flow

yatharthranjan committed:
Initial kotlin convert of added classes

yatharthranjan committed:
Rename .java to .kt

yatharthranjan committed:
various fixes and improvements

yatharthranjan committed:
Final improvements and fixes

- Add realtime consumer to docker builds for integration tests
- Fix starting for docker application when no schemas in schema registry
- Updates classes and code based on expected API spec of appserver
- Reuse classes wherever possible
- Fix timezone conversions and add jitter and tolerance to the time calculations
- Simplify configs and string formatting
- Handle unhandled exceptions

yatharthranjan committed:
Minor updates

yatharthranjan committed:
Add localisation support for AppServerIntervention.kt

- Add more properties and fixes
- update fields and time calculations based on app expectations

yatharthranjan committed:
fix referenceTimestamp

yatharthranjan committed:
exclude clinicalProtocol if null

yatharthranjan committed:
fix eval

yatharthranjan committed:
fix and test email action

- Add smtp to docker-compose.yml
- Also catch all exceptions from action execution, so other actions can proceed.

yatharthranjan committed:
Notify admin in case of action, condition or consumer failure

yatharthranjan committed:
Add docs on realtime inference consumers

yatharthranjan committed:
Merge pull request #92 from RADAR-base/finalise_realtime_inference

Finalise realtime inference

yatharthranjan committed:
JsonPathCondition.kt: return false if record value is null

yatharthranjan committed:
Add support for sending both notification and data message simultaneously

- Make delay configurable

yatharthranjan committed:
send data messages with HIGH priority

yatharthranjan committed:
make jitter configurable and null by default

- This is to prevent duplicate notifications

yatharthranjan committed:
Fix ttl

yatharthranjan committed:
add explicit "type" to protocol

yatharthranjan committed:
Add action for adding triggered task to user's schedule on appserver

yatharthranjan commented:
@mpgxvii, is it possible for you to review this when you have some free bandwidth? I think you plan to use this in the near future, so might be good to review and familiarise, unless you plan to use something else?

yatharthranjan committed:
Merge branch 'dev' into add_realtime_consumer
